src/iterative_ensemble_smoother/esmda.py

```diff
@@ -34,9 +35,96 @@
+        if covariance.ndim == 2 and covariance.shape[0] != covariance.shape[1]:
+            raise ValueError("Argument `covariance` must be square if it's 2D.")
+
+        if not (isinstance(observations, np.ndarray) and observations.ndim == 1):
+            raise TypeError("Argument `observations` must be a 1D NumPy array.")
+
+        if not observations.shape[0] == covariance.shape[0]:
+            raise ValueError("Shapes of `observations` and `covariance` must match.")
+
+        if not (
+            isinstance(seed, (int, np.random._generator.Generator)) or seed is None
+        ):
+            raise TypeError(
+                "Argument `seed` must be an integer "
+                "or numpy.random._generator.Generator."
+            )
+
+        # Store data
+        self.observations = observations
+        self.iteration = 0
+        self.rng = np.random.default_rng(seed)
+
+        # Only compute the covariance factorization once
+        # If it's a full matrix, we gain speedup by only computing cholesky once
+        # If it's a diagonal, we gain speedup by never having to compute cholesky
+        if isinstance(covariance, np.ndarray) and covariance.ndim == 2:
+            self.C_D_L = sp.linalg.cholesky(covariance, lower=False)
+        elif isinstance(covariance, np.ndarray) and covariance.ndim == 1:
+            self.C_D_L = np.sqrt(covariance)
+        else:
+            raise TypeError("Argument `covariance` must be 1D or 2D array")
+
+        self.C_D = covariance
+        assert isinstance(self.C_D, np.ndarray) and self.C_D.ndim in (1, 2)
+
+    def perturb_observations(
+        self, *, ensemble_size: int, alpha: float
+    ) -> npt.NDArray[np.double]:
+        """Create a matrix D with perturbed observations.
+
+        In the Emerick (2013) paper, the matrix D is defined in section 6.
+        See section 2(b) of the ES-MDA algorithm in the paper.
+
+        Parameters
+        ----------
+        ensemble_size : int
+            The ensemble size, i.e., the number of columns in the returned array,
+            which is of shape (num_observations, ensemble_size).
+        alpha : float
+            The covariance inflation factor. The sequence of alphas should
+            obey the equation sum_i (1/alpha_i) = 1. However, this is NOT enforced
+            in this method call. The user/caller is responsible for this.
+
+        Returns
+        -------
+        D : np.ndarray
+            Each column consists of perturbed observations, scaled by alpha.
+
+        """
+        # Draw samples from zero-centered multivariate normal with cov=alpha * C_D,
+        # and add them to the observations. Notice that
+        # if C_D = L @ L.T by the cholesky factorization, then drawing y from
+        # a zero cented normal means that y := L @ z, where z ~ norm(0, 1).
+        # Therefore, scaling C_D by alpha is equivalent to scaling L with sqrt(alpha).
+
+        D: npt.NDArray[np.double] = self.observations[:, None] + np.sqrt(
```

Comment:
We are using `np.newaxis` in other places in the code so I suggest we stick to that.
Also, this allows me to try out the suggestion-feature in GitHub 😄 

```suggestion
        D: npt.NDArray[np.double] = self.observations[:, np.newaxis] + np.sqrt(
```